pkg/controller/directimagestreammigration/trace.go

```diff
@@ -0,0 +1,74 @@
+/*
+Copyright 2020 Red Hat Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package directimagestreammigration
+
+import (
+	liberr "github.com/konveyor/controller/pkg/error"
+	migapi "github.com/konveyor/mig-controller/pkg/apis/migration/v1alpha1"
+	"github.com/konveyor/mig-controller/pkg/errorutil"
+	"github.com/konveyor/mig-controller/pkg/settings"
+	migtrace "github.com/konveyor/mig-controller/pkg/tracing"
+	"github.com/opentracing/opentracing-go"
+	"k8s.io/apimachinery/pkg/api/errors"
+)
+
+func (r *ReconcileDirectImageStreamMigration) initTracer(dism migapi.DirectImageStreamMigration) (opentracing.Span, error) {
+	// Exit if tracing disabled
+	if !settings.Settings.JaegerOpts.Enabled {
+		return nil, nil
+	}
+
+	// Set tracer on reconciler if it's not already present.
+	// We will never close this, so the 'closer' is discarded.
+	if r.tracer == nil {
+		r.tracer, _ = migtrace.InitJaeger("DirectImageStreamMigration")
+	}
+
+	// Go from dism -> dim -> migration, use migration UID to get span
+	dim, err := dism.GetOwner(r)
```

Comment:
Can we not add the migration as an owner of dism? it probably makes sense in this case if we have the ability to adopt a resource.

Say a DIM is accidently deleted, the DISM's then in this case would stick around and get re-adopted when the migration controller re-creates the DIM.

Reply:
Currently the migration owns the dim, and the dim owns the dism. I guess we could add the migration as a second owner of the dism, although I haven't really experimented with two-owner resource behavior. If there are 2 owners and one gets deleted, what happens to the resource? Also, if the DIM gets deleted, I think we'd *want* any DISMs it creates to be deleted as well. Also, the DIM controller doesn't currently have any code which would adopt pre-existing DISM resources. If there is no DISM owned by *this* DIM, then it creates one.

Reply:
One thing to keep in mind here is that it's possible for there to be a DIM that has no owner references. That just means that the user created the DIM directly rather than a migmigration creating it, which is essentially an images-only migration. Likewise with the DISM. It's possible for a single DISM to be created without an owning DIM, which will migrate a single imagestream.

Reply:
I see, we may just want to create a follow-up item to handle the re-adoption.

Reply:
OK, so first, to clarify, a resource is only deleted if the *last* owner ref is deleted, not the first, right?

And, assuming this, if we adoped the approach of adding both `MigMigration` and `DIM` owner refs, then yes, when we added MigMigration owners to DISM resources, we'd need to modify the DIM controller's DISM creation code as follows:
If no DISM owned by this DIM is found, instead of immediately moving to creation of a new DISM, we'd have a second check looking for a DISM owned by the MigMIgration owner of this DIM (if it's owned by a MigMIgration), and if it's found, then we'd adopt the existing DISM by adding the current DIM as a second owner. If now found, follow through to the existing "create a new DISM" code.

Reply:
Sounds like a good follow-up item if this is something we're interested in changing. That said, it costs us next to nothing to traverse multiple levels here since all of these resources are coming from cache. The lookups are almost instant.